File: examples/latest/navigate/android/Rerouting/app/src/main/java/com/here/rerouting/MainActivity.java
/*
 * Copyright (C) 2019-2024 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */

package com.here.rerouting;

import android.content.Context;
import android.graphics.Bitmap;
import android.os.Bundle;
import android.util.Log;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import com.here.sdk.core.engine.LogControl;
import com.here.sdk.core.engine.SDKNativeEngine;
import com.here.sdk.core.engine.SDKOptions;
import com.here.sdk.core.errors.InstantiationErrorException;
import com.here.sdk.mapview.MapScheme;
import com.here.sdk.mapview.MapView;
import com.here.sdk.routing.ManeuverAction;

public class MainActivity extends AppCompatActivity {

    private static final String TAG = MainActivity.class.getSimpleName();

    private PermissionsRequestor permissionsRequestor;
    private MapView mapView;
    private ReroutingExample reroutingExample;
    private ManeuverView maneuverView;
    private ManeuverIconProvider maneuverIconProvider;

    public interface UICallback {
        void onManeuverEvent(ManeuverAction action, String message1, String message2);
        void onRoadShieldEvent(Bitmap maneuverIcon);
        void onHideRoadShieldIcon();
        void onHideManeuverPanel();
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        // Usually, you need to initialize the HERE SDK only once during the lifetime of an application.
        initializeHERESDK();

        setContentView(R.layout.activity_main);

        // Get a MapView instance from layout.
        mapView = findViewById(R.id.map_view);
        mapView.onCreate(savedInstanceState);

        maneuverView = findViewById(R.id.maneuver_panel);
        maneuverIconProvider = new ManeuverIconProvider();
        maneuverIconProvider.loadManeuverIcons();

        handleAndroidPermissions();
    }

    private void initializeHERESDK() {
        // Disable any logs from HERE SDK.
        // Make sure to call this before initializing the HERE SDK.
        LogControl.disableLoggingToConsole();

        // Set your credentials for the HERE SDK.
        String accessKeyID = "YOUR_ACCESS_KEY_ID";
        String accessKeySecret = "YOUR_ACCESS_KEY_SECRET";
        SDKOptions options = new SDKOptions(accessKeyID, accessKeySecret);
        try {
            Context context = this;
            SDKNativeEngine.makeSharedInstance(context, options);
        } catch (InstantiationErrorException e) {
            throw new RuntimeException("Initialization of HERE SDK failed: " + e.error.name());
        }
    }

    private void handleAndroidPermissions() {
        permissionsRequestor = new PermissionsRequestor(this);
        permissionsRequestor.request(new PermissionsRequestor.ResultListener(){

            @Override
            public void permissionsGranted() {
                loadMapScene();
            }

            @Override
            public void permissionsDenied() {
                Log.e(TAG, "Permissions denied by user.");
            }
        });
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        permissionsRequestor.onRequestPermissionsResult(requestCode, grantResults);
    }

    private void loadMapScene() {
        mapView.getMapScene().loadScene(MapScheme.NORMAL_DAY, mapError -> {
            if (mapError == null) {
                reroutingExample = new ReroutingExample(MainActivity.this, mapView);
                setupEventHandling();
            } else {
                Log.d(TAG, "Loading map failed: mapErrorCode: " + mapError.name());
            }
        });
    }

    // Allow simple communication with the example class and update our UI based
    // on the events we get from the visual navigator.
    private void setupEventHandling() {
        reroutingExample.setUICallback(new UICallback() {
            @Override
            public void onManeuverEvent(ManeuverAction action, String message1, String message2) {
                Bitmap maneuverIcon = maneuverIconProvider.getManeuverIcon(action);
                String maneuverIconText = action.name();
                maneuverView.onManeuverEvent(maneuverIcon, maneuverIconText, message1, message2);
            }

            @Override
            public void onRoadShieldEvent(Bitmap maneuverIcon) {
                maneuverView.onRoadShieldEvent(maneuverIcon);
            }

            @Override
            public void onHideRoadShieldIcon() {
                maneuverView.onHideRoadShieldIcon();
            }

            @Override
            public void onHideManeuverPanel() {
                maneuverView.onHideManeuverPanel();
            }
        });
    }

    public void onShowRouteButtonClicked(View view) {
        reroutingExample.onShowRouteButtonClicked();
    }

    public void onStartStopButtonClicked(View view) {
        reroutingExample.onStartStopButtonClicked();
    }

    public void onClearMapButtonClicked(View view) {
        reroutingExample.onClearMapButtonClicked();
    }

    public void onDeviationPointsButtonClicked(View view) {
        reroutingExample.onDefineDeviationPointsButtonClicked();
    }

    public void addSpeedButtonClicked(View view) {
        reroutingExample.onSpeedButtonClicked();
    }

    @Override
    protected void onPause() {
        mapView.onPause();
        super.onPause();
    }

    @Override
    protected void onResume() {
        mapView.onResume();
        super.onResume();
    }

    @Override
    protected void onDestroy() {
        mapView.onDestroy();
        disposeHERESDK();
        super.onDestroy();
    }

    @Override
    protected void onSaveInstanceState(@NonNull Bundle outState) {
        mapView.onSaveInstanceState(outState);
        super.onSaveInstanceState(outState);
    }

    private void disposeHERESDK() {
        // Free HERE SDK resources before the application shuts down.
        // Usually, this should be called only on application termination.
        // Afterwards, the HERE SDK is no longer usable unless it is initialized again.
        SDKNativeEngine sdkNativeEngine = SDKNativeEngine.getSharedInstance();
        if (sdkNativeEngine != null) {
            sdkNativeEngine.dispose();
            // For safety reasons, we explicitly set the shared instance to null to avoid situations,
            // where a disposed instance is accidentally reused.
            SDKNativeEngine.setSharedInstance(null);
        }
    }
}

File: examples/latest/navigate/android/Kotlin/ReroutingKotlin/app/src/main/java/com/here/reroutingkotlin/MainActivity.kt
/*
 * Copyright (C) 2019-2026 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */
package com.here.reroutingkotlin

import android.os.Bundle
import android.util.Log
import android.graphics.Bitmap
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.activity.enableEdgeToEdge
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import androidx.compose.ui.viewinterop.AndroidView
import com.here.reroutingkotlin.ui.ManeuverView
import com.here.reroutingkotlin.utils.ManeuverIconProvider
import com.here.reroutingkotlin.ui.theme.ReroutingTheme
import com.here.sdk.core.engine.*
import com.here.sdk.core.errors.InstantiationErrorException
import com.here.sdk.mapview.MapFeatureModes
import com.here.sdk.mapview.MapFeatures
import com.here.sdk.mapview.MapScheme
import com.here.sdk.mapview.MapView
import com.here.sdk.routing.ManeuverAction
import com.here.sdk.units.core.utils.EnvironmentLogger
import com.here.sdk.units.core.utils.PermissionsRequestor

class MainActivity : ComponentActivity() {


    private val environmentLogger = EnvironmentLogger()
    private lateinit var permissionsRequestor: PermissionsRequestor
    private var mapView: MapView? = null
    private var reroutingExample: ReroutingExample? = null
    private var maneuverView: ManeuverView? = null
    private var maneuverIconProvider: ManeuverIconProvider? = null

    interface UICallback {
        fun onManeuverEvent(action: ManeuverAction, message1: String, message2: String)
        fun onRoadShieldEvent(maneuverIcon: Bitmap?)
        fun onHideRoadShieldIcon()
        fun onHideManeuverPanel()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        // Needs to be called before the activity is started.

        // Log application and device details.
        // It expects a string parameter that describes the application source language.
        environmentLogger.logEnvironment("Kotlin")
        permissionsRequestor = PermissionsRequestor(this)

        // Usually, you need to initialize the HERE SDK only once during the lifetime of an application.
        // Before creating a MapView instance please make sure that the HERE SDK is initialized.
        initializeHERESDK()

        enableEdgeToEdge()

        setContent {
            ReroutingTheme {
                Scaffold(
                    modifier = Modifier.fillMaxSize()
                ) { paddingValues ->
                    Box(modifier = Modifier.padding(paddingValues)) {
                        HereMapView(savedInstanceState)
                        ButtonRows()
                        ManeuverPanel()
                    }
                }
            }
        }
    }

    // Wrap the MapView into a Composable in order to use it with Jetpack Compose.
    @Composable
    private fun HereMapView(savedInstanceState: Bundle?) {
        AndroidView(factory = { context ->
            MapView(context).apply {
                mapView = this
                mapView?.onCreate(savedInstanceState)

                // Note that for this app handling of permissions is optional as no sensitive permissions
                // are required.
                // Only after permissions have been granted (if any), we load the map view and start the app.
                handleAndroidPermissions()
            }
        })
    }

    @Composable
    fun ButtonRows() {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(16.dp),
            verticalArrangement = Arrangement.spacedBy(16.dp),
            horizontalAlignment = Alignment.CenterHorizontally
        ) {
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.Center
            ) {
                CustomButton(onClick = { reroutingExample?.onShowRouteButtonClicked() }, text = "Show Route")
                Spacer(modifier = Modifier.width(16.dp))
                CustomButton(onClick = { reroutingExample?.onStartStopButtonClicked() }, text = "Start/Stop")
                Spacer(modifier = Modifier.width(16.dp))
                CustomButton(onClick = { reroutingExample?.onClearMapButtonClicked() }, text = "Clear")
            }
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.Center
            ) {
                CustomButton(onClick = { reroutingExample?.onDefineDeviationPointsButtonClicked() }, text = "Deviation Points")
                Spacer(modifier = Modifier.width(16.dp))
                CustomButton(onClick = { reroutingExample?.onSpeedButtonClicked() }, text = "Toggle Speed")
            }
        }
    }

    @Composable
    fun ManeuverPanel() {
        AndroidView(factory = { context ->
            ManeuverView(context).apply {
                maneuverView = this
            }
        }, update = { view ->
            view.redraw()
        })
    }

    @Composable
    fun CustomButton(onClick: () -> Unit, text: String) {
        Button(
            onClick = onClick,
            colors = ButtonDefaults.buttonColors(
                containerColor = Color(0xFF005CB9)
            )
        ) {
            Text(text)
        }
    }

    private fun initializeHERESDK() {
        // Disable any logs from HERE SDK. Call this before initializing the HERE SDK.
        LogControl.disableLoggingToConsole()
        // Set your credentials for the HERE SDK.
        val accessKeyID = "YOUR_ACCESS_KEY_ID"
        val accessKeySecret = "YOUR_ACCESS_KEY_SECRET"
        val authenticationMode = AuthenticationMode.withKeySecret(accessKeyID, accessKeySecret)
        val options = SDKOptions(authenticationMode)
        try {
            val context = this
            SDKNativeEngine.makeSharedInstance(context, options)
        } catch (e: InstantiationErrorException) {
            throw RuntimeException("Initialization of HERE SDK failed: " + e.error.name)
        }
    }

    // Convenience method to check all permissions that have been added to the AndroidManifest.
    private fun handleAndroidPermissions() {
        permissionsRequestor.request(object :
            PermissionsRequestor.ResultListener {
            override fun permissionsGranted() {
                loadMapScene()
            }

            override fun permissionsDenied() {
                Log.e(TAG, "Permissions denied by user.")
            }
        })
    }

    private fun loadMapScene() {
        val mapViewNonNull = mapView ?: run {
            Log.e(TAG, "mapView is null. Cannot load map scene.")
            return
        }

        // Load a scene from the HERE SDK to render the map with a map scheme.
        mapViewNonNull.mapScene.loadScene(MapScheme.NORMAL_DAY) { mapError ->
            if (mapError == null) {
                val mapFeatures = HashMap<String, String>()
                mapFeatures[MapFeatures.LOW_SPEED_ZONES] = MapFeatureModes.LOW_SPEED_ZONES_ALL
                mapViewNonNull.mapScene.enableFeatures(mapFeatures)

                maneuverIconProvider = ManeuverIconProvider().also { it.loadManeuverIcons() }
                reroutingExample = ReroutingExample(this@MainActivity, mapViewNonNull)
                setupEventHandling()
            } else {
                Log.d(TAG, "Map loading failed: ${mapError.name}")
            }
        }
    }

    private fun setupEventHandling() {
        reroutingExample?.setUICallback(object : UICallback {
            override fun onManeuverEvent(action: ManeuverAction, message1: String, message2: String) {
                val maneuverIcon = maneuverIconProvider?.getManeuverIcon(action)
                val maneuverIconText = action.name
                maneuverView?.onManeuverEvent(maneuverIcon, maneuverIconText, message1, message2)
            }

            override fun onRoadShieldEvent(maneuverIcon: Bitmap?) {
                maneuverView?.onRoadShieldEvent(maneuverIcon)
            }

            override fun onHideRoadShieldIcon() {
                maneuverView?.onHideRoadShieldIcon()
            }

            override fun onHideManeuverPanel() {
                maneuverView?.onHideManeuverPanel()
            }
        })
    }

    override fun onPause() {
        mapView?.onPause()
        super.onPause()
    }

    override fun onResume() {
        mapView?.onResume()
        super.onResume()
    }

    override fun onDestroy() {
        mapView?.onDestroy()
        disposeHERESDK()
        super.onDestroy()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        mapView?.onSaveInstanceState(outState)
        super.onSaveInstanceState(outState)
    }

    private fun disposeHERESDK() {
        reroutingExample?.dispose()

        // Free HERE SDK resources before the application shuts down.
        // Usually, this should be called only on application termination.
        // Afterwards, the HERE SDK is no longer usable unless it is initialized again.
        SDKNativeEngine.getSharedInstance()?.dispose()
        // For safety reasons, we explicitly set the shared instance to null to avoid situations,
        // where a disposed instance is accidentally reused.
        SDKNativeEngine.setSharedInstance(null)
    }

    private companion object {
        private val TAG = MainActivity::class.java.simpleName
    }
}
